packages/react-ui-core/src/Gmap/FreeDrawLayer.js

```diff
@@ -0,0 +1,93 @@
+import React, { PureComponent } from 'react'
+import PropTypes from 'prop-types'
+import autobind from 'autobind-decorator'
+import themed from 'react-themed'
+import { EVENTS } from './utils/const'
+import { setupEvents, removeEvent } from './utils/mapEventHelpers'
+
+@themed(['Gmap_FreeDrawTootip'], { pure: true })
+export default class FreeDrawLayer extends PureComponent {
+  static propTypes ={
+    map: PropTypes.object,
+    onMapDrawStart: PropTypes.func,
+    onMapDrawEnd: PropTypes.func,
+    theme: PropTypes.object,
+  }
+
+  static defaultProps = {
+    theme: {},
+    map: {},
+  }
+
+  enableMapControls() {
+    const { map } = this.props
+    map.setOptions({
+      draggable: true,
+      zoomControl: true,
+      scrollwheel: true,
+      disableDoubleClickZoom: true,
+    })
+  }
+
+  disableMapControls() {
+    const { map } = this.props
+    map.setOptions({
+      draggable: false,
+      zoomControl: false,
+      scrollwheel: false,
+      disableDoubleClickZoom: false,
+    })
+  }
+
+  @autobind
+  enableMapDraw() {
+    const { map } = this.props
+    this.disableMapControls()
+    setupEvents(map, EVENTS, {
+      onMouseDown: () => {
+        this.drawFreeHand()
+      },
+    }, true)
```

Comment:
Does this `onMouseDown` event need to get removed at some point? What happens if you enable map draw several times, will you end up with multiple event handlers on the map?

Reply:
So we modified the `setupEvents` function to take a bool as the fourth argument, if the bool is true it will use `addListenerOnce` if false or not existent it will use `addListener`. That way we can just pass in `true` as the last argument and it will only set the listener once